src/odemis/util/peak_local_max.py

```diff
@@ -0,0 +1,482 @@
+"""
+Created on 26 Jul 2021
+
+@author: Andries Effting
+
+The functions below are copied from skimage._shared.coord and
+skimage.feature.peak, with minor modification for use in Odemis.
+
+The following copyright notice applies:
+
+Copyright (C) 2019, the scikit-image team
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are
+met:
+
+ 1. Redistributions of source code must retain the above copyright
+    notice, this list of conditions and the following disclaimer.
+ 2. Redistributions in binary form must reproduce the above copyright
+    notice, this list of conditions and the following disclaimer in
+    the documentation and/or other materials provided with the
+    distribution.
+ 3. Neither the name of skimage nor the names of its contributors may be
+    used to endorse or promote products derived from this software without
+    specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
+INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+"""
+
+from warnings import warn
+from typing import Optional, Set, Tuple, Union
+
+import numpy as np
+import scipy.ndimage as ndi
+from scipy.spatial import cKDTree, distance
+
+
+def _ensure_spacing(
+    coord: np.ndarray, spacing: float, p_norm: float, max_out: Optional[int]
+) -> np.ndarray:
+    """
+    Returns a subset of coord where a minimum spacing is guaranteed.
+
+    Parameters
+    ----------
+    coord : ndarray
+        The coordinates of the considered points.
+    spacing : float
+        The maximum allowed spacing between the points.
+    p_norm : float
+        Which Minkowski p-norm to use. Should be in the range [1, inf].
+        A finite large p may cause a ValueError if overflow can occur.
+        ``inf`` corresponds to the Chebyshev distance and 2 to the
+        Euclidean distance.
+    max_out: int
+        If not None, at most the first ``max_out`` candidates are
+        returned.
+
+    Returns
+    -------
+    output : ndarray
+        A subset of coord where a minimum spacing is guaranteed.
+
+    """
+    # Use KDtree to find the peaks that are too close to each other
+    tree = cKDTree(coord)
+
+    indices = tree.query_ball_point(coord, r=spacing, p=p_norm)
+    rejected_peaks_indices: Set[int] = set()
```

Comment:
I've never seen this notation outside the function signature. What's the advantage of using it here? I checked and it doesn't force the set to be ints.

Reply:
It's a feature added to Python 3.5, see: https://docs.python.org/3/library/typing.html
It is not enforced at run-time, so it doesn't slow down a program. In my view there are two benefits:
- you can check the expected argument and return types directly from the function declaration
- you can have your code checked by a static type checker (find bugs)

This line is a type hint to indicate to the type checker that the variable `rejected_peaks_indices` is a set of ints.

Reply:
If you have mypy activated in your IDE, then it will complain if the set doesn't contain ints (or if it's not a set).